pkg/metrics/node.go

```diff
@@ -0,0 +1,179 @@
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	mgmtcontrollers "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/wrangler/pkg/ticker"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	nodeProviderLabel  = "provider"
+	nodeClusterIDLabel = "cluster_id"
+	nodeIsActiveLabel  = "is_active"
+)
+
+var (
+	nodeLabels = []string{nodeProviderLabel, nodeClusterIDLabel, nodeIsActiveLabel}
+	numNodes   = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: "cluster_manager",
+			Name:      "nodes",
+			Help:      "Number of nodes in rancher managed clusters",
+		}, nodeLabels,
+	)
+	numCores = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: "cluster_manager",
+			Name:      "node_cores",
+			Help:      "Number of node cores in rancher managed clusters",
+		}, nodeLabels,
+	)
+)
+
+const (
+	reportInterval = time.Duration(15 * time.Second)
+	logPrefix      = "[prometheus-node-metrics]"
+)
+
+type nodeMetrics struct {
+	nodeCache    mgmtcontrollers.NodeCache
+	clusterCache mgmtcontrollers.ClusterCache
+}
+
+func (m *nodeMetrics) collect(ctx context.Context) {
+	for range ticker.Context(ctx, reportInterval) {
+		logrus.Debugf("%s collecting nodes to report metrics", logPrefix)
+
+		nodes, err := m.nodeCache.List("", labels.Everything())
+		if err != nil {
+			logrus.Errorf("%s couldn't list v3.Nodes: %v", logPrefix, err)
+			continue
+		}
+
+		var infos []*nodeInfo
+		for _, node := range nodes {
+			info, err := m.getNodeInfo(node)
+			if err != nil {
+				logrus.Debugf("%s could not determine node info: %v", logPrefix, err)
+				continue
+			}
+			infos = append(infos, info)
+		}
+
+		setMetrics(infos)
+	}
+
+	logrus.Debugf("%s context cancelled, exiting", logPrefix)
+}
+
+type nodeLabelValues struct {
+	provider  string
+	clusterID string
+	isActive  bool
+}
+
+type nodeInfo struct {
+	coreCount   int
+	labelValues nodeLabelValues
+}
+
+const (
+	unknownProvider = "unknown"
+)
+
+// getNodeInfo uses the v3.Node object and it's corresponding v3.Cluster object to get metric info for a node
+func (m *nodeMetrics) getNodeInfo(node *v3.Node) (*nodeInfo, error) {
+	cluster, err := m.clusterCache.Get(node.ObjClusterName())
+	if err != nil {
+		return nil, fmt.Errorf("error getting cluster %s for node %s, %v", node.ObjClusterName(), node.Name, err)
+	}
+	if cluster == nil {
+		return nil, fmt.Errorf("could not get associated cluster for node: %s/%s", node.Namespace, node.Name)
+	}
+	if cluster.DeletionTimestamp != nil {
+		return nil, fmt.Errorf("associated cluster for node: %s/%s is being deleted", node.Namespace, node.Name)
+	}
+
+	// a node is considered to be active if it is ready and it belongs to a ready cluster
+	isActive := v32.ClusterConditionReady.IsTrue(cluster) && v32.NodeConditionReady.IsTrue(node)
```

Comment:
From last session, this won't actually matter during calculations, but should be fine to present via endpoint.